feat(eslint-plugin): [no-unnecessary-type-parameters] initial implementation (#8173)
typescript-eslint/typescript-eslint: packages/eslint-plugin/docs/rules/no-unnecessary-type-parameters.md

---
description: 'Disallow type parameters that only appear once'
---

> 🛑 This file is source code, not the primary documentation location! 🛑
>
> See **https://typescript-eslint.io/rules/no-unnecessary-type-parameters** for documentation.

This rule forbids type parameters that only appear once in a function, method, or class definition.

> Type parameters relate two types. If a generic type parameter only appears once, then it is not relating anything. It can usually be replaced with explicit types such as `unknown`.

Reviewer: [Docs] Nit: "generic" in "generic type parameter" is redundant, I think. As, all type parameters are "generic".

```suggestion
Type parameters relate two types. If a type parameter only appears once, then it is not relating anything. It can usually be replaced with explicit types such as `unknown`.
```

> At best unnecessary type parameters make code harder to read. At worst they can be used to disguise unsafe type assertions.

Reviewer: [Praise] I really like this line 😄

> ## Options
> 
> None.
> 

Reviewer: [Docs] Our tooling will auto-generate this ✨  https://github.com/typescript-eslint/typescript-eslint/blob/b1c92bbca17021605950308987ebc0383037f649/packages/website/plugins/generated-rule-docs/insertions/insertNewRuleReferences.ts#L74

```suggestion
```

## Examples

<!--tabs-->

### ❌ Incorrect

```ts

> function third<A, B, C>(a: A, b: B, c: C): C {

Reviewer: [Docs] Nit: `third` is the hardest to parse out of the examples for me, while `parseYAML` is a pretty much perfect showcase of the rule. What do you think about making `third` either last or second-to-last, and or making the name slightly more descriptive - like `onlyThirdTypeParameterIsNeeded`?

  return c;
}

function parseYAML<T>(input: string): T {
  // ...
}

function printProperty<T, K extends keyof T>(obj: T, key: K) {
  console.log(obj[key]);
}

class Joiner<T extends string | number> {
  join(els: T[]) {
    return els.map(el => '' + el).join(',');
  }
}
```

### ✅ Correct

```ts
function third<A, B, C>(a: unknown, b: unknown, c: C): C {
  return c;
}

function parseYAML(input: string): unknown {
  // ...
}

function printProperty<T>(obj: T, key: keyof T) {
  console.log(obj[key]);
}

class Joiner {
  join(els: (string | number)[]) {
    return els.map(el => '' + el).join(',');
  }
}

> // T appears twice: in the type of arg and as the return type

Reviewer: [Praise] These two added valid examples are very clear & helpful for explaining the rule IMO. 🔥

function identity<T>(arg: T): T {
  return arg;
}

// T appears twice: "keyof T" and in the inferred return type (T[K]).
// K appears twice: "key: K" and in the inferred return type (T[K]).
function getProperty<T, K extends keyof T>(obj: T, key: K) {
  return obj[key];
}
```

## When Not To Use It

> This rule will prevent using type parameters to test types, for example:

Reviewer: [Docs] Nit: a bit more precision here. The rules don't _prevent_ anything, they just report on violations. Some users mistakenly think rules can't be configured or overridden.

```suggestion
This rule will report on functions that use type parameters solely to test types, for example:
```

```ts
function assertType<T>(arg: T) {}

assertType<number>(123);
assertType<number>('abc');
//                 ~~~~~
// Argument of type 'string' is not assignable to parameter of type 'number'.
```

> If you're using this pattern then you'll want to disable this rule.

Reviewer: [Docs] I think a devs who are doing this types testing pattern will want a better workaround than just not using the rule. How about giving them the following two suggestions?:

* Several popular libraries with links? I.e. https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type and the list under https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/tree/db69afcca880b434d518569ea38746ca5b2f3eb8#references ?
* Use inline disables in functions specifically used for testing
    * For reference, #7447 added details like this to many rule docs pages & somewhat standardized some common phrasing

Reviewer: Coming back to this, I think it's fine to not for now. If folks are testing types then they should already be at least a little familiar with these strategies.

> If you're using generic type parameters to implement phantom types then you'll also run afoul of this rule.

Reviewer: [Docs] I don't know what phantom types are, and neither will many readers of this. This should either define it or be omitted. A quick definition and link to further reading would be sufficient.
